scripts/stats_std.py

```diff
@@ -17,59 +17,58 @@ def mean_of_array(reads,start, pos):
 	return mean
 
 def main():
-	mybam = pysam.AlignmentFile('UMB.aligned.sorted.bam')
+	mybam = pysam.AlignmentFile('MGH283_align_sort.bam')
 
-	# iterate over statistics, one record at a time
-	#stats = pysamstats.load_coverage(mybam, chrom='NZ_CP023386.1', start= 2000000, end=3000000)
-	stats = pysamstats.load_coverage(mybam)
+	file = 'output_std_mgh_contig.txt'
 
-	reads = stats.reads_all
-	position = stats.pos
-	chrom = stats.chrom
+	with open(file, "w") as f:
+		print("Index Chrom Pos Thresh Diff", file=f)
+		for contig in mybam.header.references:  # Do the analysis per contig
+			stats = pysamstats.load_coverage(mybam, chrom=contig)
 
-	print(len(reads))
-	print(len(position))
-	print(len(chrom))
+			reads = stats.reads_pp
+			position = stats.pos
+			chrom = stats.chrom
 
-	thresh_up, thresh_down = threshold(reads)
-	#print(thresh_up)
-	#print(thresh_down)
+			print(len(reads))
+			print(len(position))
+			print(len(chrom))
 
-	start = 0
-	big_start = 0
-	last_end = 0
+			thresh_up, thresh_down = threshold(reads)
+			#print(thresh_up)
+			#print(thresh_down)
 
-	file = 'output_std.txt'
+			start = 0
+			big_start = 0
+			last_end = 0
 
-	with open(file, "w") as f:
-		print("Index Chrom Pos Thresh Diff", file=f)
-		for i in range(len(reads)):
-			if i == ((len(reads)) - 1):
-				current = reads[i]
-				next_one = reads[0]
-				next_pos = position[0]
-			else:
-				current = reads[i]
-				next_one = reads[i+1]
-				next_pos = position[i+1]
+			for i in range(1000, len(reads), 1000):
+				if i == ((len(reads)) - 1):
+					current = reads[i]
+					next_one = reads[0]
+					next_pos = position[0]
+				else:
+					current = reads[i]
+					next_one = reads[i+1]
+					next_pos = position[i+1]
 
-			if (i % 1000) == 0 and i != 0:
-				#print(str(i) + ' : ' + str(position[i]))
 				mean = mean_of_array(reads, start, i)
 				if mean >= thresh_up:
 					if last_end == start:
-						print(str(big_start) + ' ' + str(i) + ' ' + str(mean) + ' +3sigma ' + str(chrom[big_start]).lstrip('b\'').rstrip('\'') + ':' + str(position[big_start]) + ' ' + str(chrom[i]).lstrip('b\'').rstrip('\'') + ':' + str(position[i]), file=f)
-						print(str(big_start) + ' ' + str(i) + ' ' + str(mean) + ' +3sigma ' + str(chrom[big_start]).lstrip('b\'').rstrip('\'') + ':' + str(position[big_start]) + ' ' + str(chrom[i]).lstrip('b\'').rstrip('\'') + ':' + str(position[i]))
+						big_start_chrom = str(chrom[big_start]).lstrip('b\'').rstrip('\'')
```

Comment:
For some reason, the elements in the chromosome array created by pysamstats have quotation marks around them and a b at the start of the string (e.g. b'NZ_CP060654.1') so I had to use l and rstip to remove those. I tried to do string indexing but that didn't work. I had to convert to str first because the lstrip and rstrip functions didn't work without it being converted to a string first.

Reply:
Ahh, that's because apparently pysam returns the chromosome name as `bytes` object. You'll need to specify how Python should interpret these bytes to convert it to a string. UTF-8 is almost always the correct choice:

`chrom[big_start'].decode('utf-8')`

To learn more about `bytes` versus `str`: https://betterprogramming.pub/strings-unicode-and-bytes-in-python-3-everything-you-always-wanted-to-know-27dc02ff2686

Reply:
Ahhh ok great i'll do that. Thanks!!